saslauthenticate.go

```diff
@@ -0,0 +1,51 @@
+package kafka
+
+import (
+	"bufio"
+)
+
+type saslAuthenticateRequestV0 struct {
+	// Data holds the SASL payload
+	Data []byte
+}
+
+func (t saslAuthenticateRequestV0) size() int32 {
+	return sizeofBytes(t.Data)
+}
+
+func (t saslAuthenticateRequestV0) writeTo(w *bufio.Writer) {
+	writeBytes(w, t.Data)
+}
+
+type saslAuthenticateResponseV0 struct {
+	// ErrorCode holds response error code
+	ErrorCode int16
+
+	ErrorMessage string
+
+	Data []byte
+}
+
+func (t saslAuthenticateResponseV0) size() int32 {
+	return sizeofInt16(t.ErrorCode) + sizeofString(t.ErrorMessage) + sizeofBytes(t.Data)
+}
+
+func (t saslAuthenticateResponseV0) writeTo(w *bufio.Writer) {
+	writeInt16(w, t.ErrorCode)
+	writeString(w, t.ErrorMessage)
+	writeBytes(w, t.Data)
+}
+
+func (t *saslAuthenticateResponseV0) readFrom(r *bufio.Reader, sz int) (remain int, err error) {
+	if remain, err = readInt16(r, sz, &t.ErrorCode); err != nil {
+		return
+	}
+	// TODO: type is nullable, size = -1 means null, readString does not support this?
```

Comment:
Add readNullableString with **string?

Reply:
I believe readString does handle nullable strings in a reasonable way. We have this sequence of function calls: readString -> readStringWith -> readNewString -> readNewBytes. readNewBytes returns an empty slice when size = -1 which means that readString will return the empty string in this case. So if readString returns the empty string either the empty string was read (size = 0) or the string wasn't provided (size = -1). For this particular code I don't think there is a need to tell the two cases apart.